sssd_test_framework/utils/realmd.py

```diff
@@ -0,0 +1,117 @@
+"""Manage realm Operations."""
+
+from __future__ import annotations
+
+from pytest_mh import MultihostHost, MultihostUtility
+from pytest_mh.conn import ProcessResult
+
+__all__ = [
+    "RealmUtils",
+]
+
+
+class RealmUtils(MultihostUtility[MultihostHost]):
+    """
+    Interface to the realm utility.
+
+    .. code-block:: python
+        :caption: Example usage
+
+        @pytest.mark.topology(KnownTopology.AD)
+        def test_realm_discover(client: Client, provider: ADProvider):
+            r = client.realm.discover([p_domain, "--use-ldaps"])
+            assert provider.host.domain in r.stdout, "realm failed to discover domain info"
+
+    """
+
+    def discover(self, *, args: list[str] | None = None) -> ProcessResult:
+        """
+        Discover a realm and it's capabilities.
+
+        :param args: Additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["realm", "discover", *args])
+
+    def leave(self, args: list[str] | None = None) -> ProcessResult:
+        """
+        Deconfigure and remove a client
+
+        :param args: Additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["realm", "leave", *args])
+
+    def join(
+        self,
+        domain: str,
+        *,
+        args: list[str] | None = None,
+        password: str | None = None,
+        user: str | None = None,
+        krb: bool = False,
+    ) -> ProcessResult:
+        """
+        Join and configure a client
+
+        :param domain: Domain to join.
+        :type domain: str
+        :param args: Additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        :param password: Password to run the operation, defaults to None
+        :type password: str
+        :param krb: Enable kerberos authentication, defaults to False
+        :type krb: bool
+        """
+        if args is None:
+            args = []
+
+        if krb:
+            self.host.conn.exec(["kinit", user], input=password)
+            command = self.host.conn.exec(["realm", "join", "--verbose", *args, domain])
+        else:
+            command = self.host.conn.exec(["realm", "join", "--verbose", *args, "-U", user, domain], input=password)
+
+        return command
+
+    def rlist(self, *, args: list[str] | None = None) -> ProcessResult:
+        """
+        List discovered, and configured realms.
+
+        :param args: Additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["realm", "list", *args])
+
+    def permit(self, *, args: list[str] | None = None) -> ProcessResult:
```

Comment:
It'd be nice to use this like

client.realm.permit() # realm permit -a 
client.realm.permit(user="user1" | user=["user1", "user2"])
client.realm.permit(group="group")

Reply:
Need to evaluate

Reply:
Like join, but this is has an extra step. 

```
def permit(self, *, user: str | list[str] | None = None, group:  str  | list[str] | None = None,  args: list[str] | None = None) -> ProcessResult: 
```

user parameter takes type str, list[str] or None, and None = None, one extra step, check if it's a string or not. 

```
user = *user if isinstance(user, list) else user = ""
```
is the same as 
```
if isinstance(user, list): 
    user = *user
if user is None: 
    user = ""
```

Since user is a string, it will never hit the condition and just the desired type.  If it's a list, turn it into a string, anything else it's an empty string.

Reply:
dropping 'permit' method for now.